File: LeetSync/3219-make-lexicographically-smallest-array-by-swapping-elements/make-lexicographically-smallest-array-by-swapping-elements.py
class Solution:
    def lexicographicallySmallestArray(self, nums: List[int], limit: int) -> List[int]:
        n = len(nums)

        sorted_nums = sorted(nums)

        group_queues = []
        num_to_group = {}

        for num in sorted_nums:
            if not group_queues or num - group_queues[-1][-1] > limit:
                group_queues.append(deque())

            group_queues[-1].append(num)
            num_to_group[num] = len(group_queues) - 1

        result = [0] * n
        for i, num in enumerate(nums):
            group = num_to_group[num]
            result[i] = group_queues[group].popleft()

        return result

File: LeetSync/3219-make-lexicographically-smallest-array-by-swapping-elements/make-lexicographically-smallest-array-by-swapping-elements.cpp
class Solution {
public:
    vector<int> lexicographicallySmallestArray(vector<int>& nums, int limit) {
        int n = nums.size();
        
        vector<int> sortedNums = nums;
        sort(sortedNums.begin(), sortedNums.end());

        vector<std::deque<int>> groupQueues;
        unordered_map<int, int> numToGroup;

        for (int i = 0; i < n; ++i) {
            if (i == 0 || sortedNums[i] - sortedNums[i - 1] > limit) {
                groupQueues.push_back({});
            }
            groupQueues.back().push_back(sortedNums[i]);
            numToGroup[sortedNums[i]] = groupQueues.size() - 1;
        }

        vector<int> result(n);
        for (int i = 0; i < n; ++i) {
            int group = numToGroup[nums[i]];
            result[i] = groupQueues[group].front();
            groupQueues[group].pop_front();
        }

        return result;
    }
};